lnode/synctxnpool.go

```diff
@@ -0,0 +1,349 @@
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_RPL_TXN_POOL_HASH,
+		Message:     buf,
+	}
+
+	return msg, nil
+}
+
+// build REQ_SYNC_TXN_POOL request message. This message includse my txn pool's txn address and maximum nonce.
+func NewReqSyncTxnPoolMessage(tp *pool.TxnPool) (*pb.UnsignedMessage, error) {
+
+	addrNonceList := make([]*pb.AddrNonce, 0)
+	mapAddrNonce := tp.GetTwtAddrNonce()
+
+	for addr, nonce := range mapAddrNonce {
+		addrNonceList = append(addrNonceList, &pb.AddrNonce{
+			Address: addr.ToArray(),
+			Nonce:   nonce,
+		})
+	}
+
+	// get the duration since last sync txn time.
+	duration := time.Now().UnixMilli() - tp.GetSyncTime()
+	if tp.GetSyncTime() == 0 { // never synced before
+		duration = 0
+	}
+	msgBody := &pb.RequestSyncTxnPool{Duration: duration, AddrNonce: addrNonceList}
+
+	buf, err := proto.Marshal(msgBody)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_REQ_SYNC_TXN_POOL,
+		Message:     buf,
+	}
+
+	return msg, nil
+}
+
+// build RPL_SYNC_TXN_POOL message. reply transactions which are needed to sync to the requestor.
+func NewRplSyncTxnPoolMessage(reqMsg *pb.RequestSyncTxnPool, tp *pool.TxnPool) (*pb.UnsignedMessage, error) {
+
+	mapReqAddrNonce := make(map[common.Uint160]uint64)
+
+	// if there is address and nonce in the request message, save them in a map for better searching.
+	if len(reqMsg.AddrNonce) > 0 {
+		for _, addrNonce := range reqMsg.AddrNonce {
+			mapReqAddrNonce[common.BytesToUint160(addrNonce.Address)] = addrNonce.Nonce
+		}
+	}
+
+	var respTxns []*pb.Transaction
+	var err error
+
+	if reqMsg.Duration > 0 {
+		respTxns, err = syncInDuration(tp, mapReqAddrNonce, reqMsg.Duration)
+	} else {
+		respTxns, err = syncFully(tp, mapReqAddrNonce)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	msgBody := &pb.Transactions{Transactions: respTxns}
+
+	buf, err := proto.Marshal(msgBody)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_RPL_SYNC_TXN_POOL,
+		Message:     buf,
+	}
+
+	return msg, nil
+}
+
+// Only sync txn in the last duration seconds, partially of the txn pool
+func syncInDuration(tp *pool.TxnPool, reqAddrNonce map[common.Uint160]uint64, duration int64) ([]*pb.Transaction, error) {
+
+	now := time.Now().UnixMilli()
+	earliest := now - duration - 1000 // push 1 second ahead
+	respTxns, err := tp.GetTwtTxnAfter(earliest, reqAddrNonce)
+
+	return respTxns, err
+}
+
+// sync all my txns in the pool to the neighbor when the neighbor just start.
+func syncFully(tp *pool.TxnPool, reqAddrNonce map[common.Uint160]uint64) ([]*pb.Transaction, error) {
+
+	respTxns := make([]*pb.Transaction, 0) // response txn list buffer
+
+	mapMyTxnList := tp.GetAllTransactionLists()
+	for addr, myTxnList := range mapMyTxnList {
+		nonce, ok := reqAddrNonce[addr]
+		if ok { // if addr is in request list, we append bigger nonce txns to response list
```

Comment:
No need for this if. `nonce` is 0 if `!ok`

Reply:
Yes, modified.